### PR TITLE
Add check_connectivity and endpoints methods to code generated service.

### PR DESCRIPTION
**Pull request by fmassot**
New things:
- use `SocketAddr` as the balance channel key in the generated client. This is mostly for convenience and to insure a consistent usage of the clients.
- Add `check_connectivity` method to generated clients. Note that it will return always OK if the channel is pure tonic channel which is a bit strange. We need to add some warning here until we fix this issue or fix it in this PR. 
- add `endpoints` to generated clients.
- Remove the usage of tower `Timeout` as tonic already set a timeout layer on the grpc client (because we set the timeout of the `Endpoint`).
- Update the balance channel to keep track of available keys and be able to retrieve them.

### CODE REVIEW

File: quickwit/quickwit-codegen/example/src/codegen/hello.rs [hunk 1]
```diff
@@ -102,9 +104,22 @@ impl HelloClient {
...
-            HelloGrpcClientAdapter::new(hello_grpc_client::HelloGrpcClient::new(channel)),
-        )
+        let (_, num_connections_watcher) = tokio::sync::watch::channel(1);
+        let adapter = HelloGrpcClientAdapter::new(
+            hello_grpc_client::HelloGrpcClient::new(channel),
+            num_connections_watcher,
+        );
+        Self::new(adapter)
+    }
+    pub fn from_balanced_channel<K: std::hash::Hash + Eq + Send + Clone + 'static>(
```
Comment (fmassot) [inline quickwit/quickwit-codegen/example/src/codegen/hello.rs:114]: I would keep only the `from_balanced_channel`.

File: quickwit/quickwit-codegen/src/codegen.rs [hunk 2]
```diff
@@ -412,7 +484,16 @@ fn generate_client(context: &CodegenContext) -> TokenStream {
                         Output = Result<http::Response<hyper::Body>, quickwit_common::tower::BoxError>,
                     > + Send + 'static,
             {
-                #client_name::new(#grpc_client_adapter_name::new(#grpc_client_package_name::#grpc_client_name::new(channel)))
+                let (_, num_connections_watcher) = tokio::sync::watch::channel(1);
+                let adapter = #grpc_client_adapter_name::new(#grpc_client_package_name::#grpc_client_name::new(channel), num_connections_watcher);
+                Self::new(adapter)
+            }
+
+            pub fn from_balanced_channel<K: std::hash::Hash + Eq + Send + Clone + 'static>(balanced_channel: quickwit_common::tower::BalanceChannel<K>) -> #client_name
```
Comment (fmassot) [inline quickwit/quickwit-codegen/src/codegen.rs:492]: I would keep only the from_balanced_channel.

File: quickwit/quickwit-codegen/src/codegen.rs [hunk 3]
```diff
@@ -779,6 +876,22 @@ fn generate_tower_mailbox(context: &CodegenContext) -> TokenStream {
...
+        quote! {
+            async fn check_connectivity(&mut self) -> anyhow::Result<()> {
+                if self.inner.is_disconnected() {
+                    anyhow::bail!("Mailbox of actor `{}` is disconnected", self.inner.actor_instance_id())
+                }
+                Ok(())
+            }
+
+            fn uris(&self) -> Vec<quickwit_common::uri::Uri> {
+                vec![quickwit_common::uri::Uri::from_well_formed(&format!("mailbox://localhost/{}", self.inner.actor_instance_id()))]
```
Comment (fmassot) [inline quickwit/quickwit-codegen/src/codegen.rs:889]: fancy uri for actor...

File: quickwit/quickwit-codegen/example/src/codegen/hello.rs [hunk 4]
```diff
@@ -604,6 +670,18 @@ where
...
+    async fn check_connectivity(&mut self) -> anyhow::Result<()> {
+        if *self.num_connections_rx.borrow() == 0 {
+            anyhow::bail!("No connection to the server")
+        }
+        Ok(())
+    }
+    fn uris(&self) -> Vec<quickwit_common::uri::Uri> {
+        vec![
+            quickwit_common::uri::Uri::from_well_formed(&
+            format!("grpc://{}.service.cluster", "Hello"))
```
Comment (guilload) [inline quickwit/quickwit-codegen/example/src/codegen/hello.rs:682]: Ideally, there should be as many URIs as connections in the pool. Each URI should be `grpc://<host:port>/<proto package name>.<proto service name>`.

File: quickwit/quickwit-codegen/example/src/codegen/hello.rs [hunk 5]
```diff
@@ -604,6 +670,18 @@ where
             })
             .map_err(|error| error.into())
     }
+    async fn check_connectivity(&mut self) -> anyhow::Result<()> {
+        if *self.num_connections_rx.borrow() == 0 {
+            anyhow::bail!("No connection to the server")
+        }
+        Ok(())
+    }
+    fn uris(&self) -> Vec<quickwit_common::uri::Uri> {
```
Comment (guilload) [inline quickwit/quickwit-codegen/example/src/codegen/hello.rs:679]: Endpoints is probably a better name.

File: quickwit/quickwit-codegen/src/codegen.rs [hunk 6]
```diff
@@ -33,13 +33,15 @@ impl Codegen {
         out_dir: &str,
         result_type_path: &str,
         error_type_path: &str,
+        additional_service_methods: bool,
```
Comment (guilload) [inline quickwit/quickwit-codegen/src/codegen.rs:36]: `generate_extra_service_methods`

File: quickwit/quickwit-codegen/src/codegen.rs [hunk 7]
```diff
@@ -897,21 +1011,43 @@ fn generate_mailbox_bounds_and_methods(
...
+            async fn check_connectivity(&mut self) -> anyhow::Result<()> {
+                if *self.num_connections_rx.borrow() == 0 {
+                    anyhow::bail!("No connection to the server")
+                }
+                Ok(())
+            }
+
+            fn uris(&self) -> Vec<quickwit_common::uri::Uri> {
+                // TODO: use real uris.
+                vec![quickwit_common::uri::Uri::from_well_formed(&format!("grpc://{}.service.cluster", #service_name_string))]
```
Comment (guilload) [inline quickwit/quickwit-codegen/src/codegen.rs:1030]: You don't need format, just generate the static string directly.

File: quickwit/quickwit-codegen/src/codegen.rs [hunk 8]
```diff
@@ -897,21 +1011,43 @@ fn generate_mailbox_bounds_and_methods(
...
+    let service_name_string = service_name.to_string();
     let grpc_client_package_name = &context.grpc_client_package_name;
     let grpc_client_name = &context.grpc_client_name;
     let grpc_client_adapter_name = &context.grpc_client_adapter_name;
     let grpc_server_adapter_methods = generate_grpc_client_adapter_methods(context);
+    let additional_grpc_server_adapter_methods = if context.additional_service_methods {
+        quote! {
+            async fn check_connectivity(&mut self) -> anyhow::Result<()> {
+                if *self.num_connections_rx.borrow() == 0 {
+                    anyhow::bail!("No connection to the server")
```
Comment (guilload) [inline quickwit/quickwit-codegen/src/codegen.rs:1023]:
```suggestion
                    anyhow::bail!("no server currently available")
```

File: quickwit/quickwit-codegen/example/src/lib.rs [hunk 9]
```diff
@@ -112,6 +113,22 @@ fn spawn_ping_response_stream(
     service_stream
 }
 
+#[async_trait]
```
Comment (guilload) [inline quickwit/quickwit-codegen/example/src/lib.rs:116]: I don't think this code and the associated test are codegen-specific.

File: quickwit/quickwit-codegen/example/src/lib.rs [hunk 10]
```diff
@@ -135,6 +152,14 @@ impl Hello for HelloImpl {
     ) -> HelloResult<HelloStream<PingResponse>> {
         Ok(spawn_ping_response_stream(request))
     }
+
+    async fn check_connectivity(&mut self) -> anyhow::Result<()> {
+        Ok(())
+    }
+
+    fn uris(&self) -> Vec<Uri> {
+        vec![]
```
Comment (guilload) [inline quickwit/quickwit-codegen/example/src/lib.rs:161]:
```suggestion
        Vec::new()
```

File: quickwit/quickwit-codegen/example/src/lib.rs [hunk 11]
```diff
@@ -301,6 +343,24 @@ mod tests {
...
             .unwrap_err();
         assert!(matches!(error, TrySendError::Closed(_)));
+
+        grpc_client.check_connectivity().await.unwrap();
+        assert_eq!(
+            grpc_client.uris(),
+            vec![Uri::from_well_formed("grpc://Hello.service.cluster")]
+        );
+
+        // Connectivity fails if there is no client.
```
Comment (guilload) [inline quickwit/quickwit-codegen/example/src/lib.rs:353]:
```suggestion
        // The connectivity check fails if there is no client behind the channel.
```

File: quickwit/quickwit-codegen/src/codegen.rs (hunk 2 above)
Comment (guilload) [reply quickwit/quickwit-codegen/src/codegen.rs:492]: Yes, let's just keep the `from_balanced_channel`. The other one is harmful.

Comment (fmassot) [reply quickwit/quickwit-codegen/src/codegen.rs]: I finally kept the `from_channel` too.

File: quickwit/quickwit-codegen/src/codegen.rs [hunk 12]
```diff
@@ -897,21 +994,47 @@ fn generate_mailbox_bounds_and_methods(
...
+                    anyhow::bail!("no server currently available")
+                }
+                Ok(())
+            }
+
+            fn endpoints(&self) -> Vec<quickwit_common::uri::Uri> {
+                self.connection_addrs_rx
+                    .borrow()
+                    .iter()
+                    .map(|addr| quickwit_common::uri::Uri::from_well_formed(format!(r"grpc://{}/{}.{}", addr, #grpc_client_package_name_string, #service_name_string)))
```
Comment (guilload) [inline quickwit/quickwit-codegen/src/codegen.rs:1016]: Just for fun:
```rust
{
    let endpoint_template = format!(r#""grpc://{{}}/{}.{}""#, context.package_name, service_name).parse::<proc_macro2::Literal>().expect("string should be a valid string literal");
    quote! {
        ...
        .map(|addr| quickwit_common::uri::Uri::from_well_formed(format!(#endpoint_template, addr)))
        ....
    }
}

**Review by guilload — APPROVED** [review]

**Pull request merged**